File: android/src/main/java/com/tonyxlh/capacitor/camera/TextRecognitionBlurHelper.java
package com.tonyxlh.capacitor.camera;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.Log;

import com.google.mlkit.vision.common.InputImage;
import com.google.mlkit.vision.text.Text;
import com.google.mlkit.vision.text.TextRecognition;
import com.google.mlkit.vision.text.TextRecognizer;
import com.google.mlkit.vision.text.latin.TextRecognizerOptions;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.CountDownLatch;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicReference;

/**
 * Text Recognition Blur Detection Helper using Google ML Kit
 * Detects blur/sharp images based on text readability and confidence scores
 */
public class TextRecognitionBlurHelper {
    private static final String TAG = "TextRecognitionBlurHelper";
    private static final double MIN_WORD_CONFIDENCE = 0.8; // 80% confidence threshold
    private static final int TIMEOUT_MS = 5000; // 5 second timeout for text recognition
    private static final double AT_LEAST_N_PERCENT_OF_WORDS_ARE_READABLE = 0.6; // 60% of words are readable
    private static final double AT_LEAST_N_PERCENT_OF_AVERAGE_CONFIDENCE = 0.85; // 85% of average confidence
    
    private TextRecognizer textRecognizer;
    private boolean isInitialized = false;
    
    // Optional dictionary for enhanced readability check
    private Set<String> commonWords;
    private boolean useDictionaryCheck = false;

    public TextRecognitionBlurHelper() {
        // Initialize common English words for dictionary check
        initializeCommonWords();
    }

    /**
     * Initialize the text recognizer
     * @param context Android context
     * @return true if initialization successful
     */
    public boolean initialize(Context context) {
        try {
            textRecognizer = TextRecognition.getClient(TextRecognizerOptions.DEFAULT_OPTIONS);
            isInitialized = true;
            Log.d(TAG, "Text recognizer initialized successfully");
            return true;
        } catch (Exception e) {
            Log.e(TAG, "Failed to initialize text recognizer: " + e.getMessage(), e);
            isInitialized = false;
            return false;
        }
    }

    /**
     * Detect if image is blurry based on text recognition confidence
     * @param bitmap Input image bitmap
     * @return true if image is blurry (text not readable), false if sharp
     */
    public boolean isBlurry(Bitmap bitmap) {
        if (!isInitialized || textRecognizer == null) {
            Log.w(TAG, "Text recognizer not initialized");
            return false; // Default to not blurry if not initialized
        }

        Log.d(TAG, "Starting text recognition blur detection...");
        try {
            TextRecognitionResult result = detectTextWithConfidence(bitmap);
            Log.d(TAG, String.format("Text recognition result: %s (confidence: %.3f, words: %d/%d)", 
                    result.isReadable ? "READABLE" : "BLURRY", 
                    result.averageConfidence, result.readableWords, result.totalWords));
            
            boolean isBlurry = !result.isReadable;
            Log.d(TAG, String.format("Blur detection result: %s (returning %.1f)", 
                    isBlurry ? "BLURRY" : "SHARP", isBlurry ? 1.0 : 0.0));
            
            return isBlurry;
        } catch (Exception e) {
            Log.e(TAG, "Error during text recognition: " + e.getMessage(), e);
            return false; // Default to not blurry on error
        }
    }

    /**
     * Detect blur with detailed confidence scores
     * @param bitmap Input image bitmap
     * @return Map with isBlur, textConfidence, wordCount, and readableWords
     */
    public java.util.Map<String, Object> detectBlurWithConfidence(Bitmap bitmap) {
        java.util.Map<String, Object> result = new java.util.HashMap<>();
        
        if (!isInitialized || textRecognizer == null) {
            Log.w(TAG, "Text recognizer not initialized");
            result.put("isBlur", false);
            result.put("textConfidence", 0.0);
            result.put("wordCount", 0);
            result.put("readableWords", 0);
            result.put("error", "Text recognizer not initialized");
            return result;
        }

        try {
            TextRecognitionResult recognitionResult = detectTextWithConfidence(bitmap);
            
            result.put("isBlur", !recognitionResult.isReadable);
            result.put("textConfidence", recognitionResult.averageConfidence);
            result.put("wordCount", recognitionResult.totalWords);
            result.put("readableWords", recognitionResult.readableWords);
            result.put("hasText", recognitionResult.totalWords > 0);
            
            Log.d(TAG, String.format("Text Recognition Result - Readable: %s, Confidence: %.3f, Words: %d/%d",
                    recognitionResult.isReadable, recognitionResult.averageConfidence, 
                    recognitionResult.readableWords, recognitionResult.totalWords));
            
            return result;
            
        } catch (Exception e) {
            Log.e(TAG, "Error during text recognition: " + e.getMessage(), e);
            result.put("isBlur", false);
            result.put("textConfidence", 0.0);
            result.put("wordCount", 0);
            result.put("readableWords", 0);
            result.put("error", e.getMessage());
            return result;
        }
    }

    /**
     * Detect text in image with confidence analysis
     * @param bitmap Input image bitmap
     * @return TextRecognitionResult with confidence and readability info
     */
    private TextRecognitionResult detectTextWithConfidence(Bitmap bitmap) {
        InputImage image = InputImage.fromBitmap(bitmap, 0);
        
        // Use CountDownLatch for synchronous operation
        CountDownLatch latch = new CountDownLatch(1);
        AtomicReference<TextRecognitionResult> resultRef = new AtomicReference<>();
        AtomicBoolean hasError = new AtomicBoolean(false);
        
        textRecognizer.process(image)
            .addOnSuccessListener(visionText -> {
                try {
                    TextRecognitionResult result = analyzeTextConfidence(visionText);
                    resultRef.set(result);
                } catch (Exception e) {
                    Log.e(TAG, "Error analyzing text confidence: " + e.getMessage(), e);
                    hasError.set(true);
                } finally {
                    latch.countDown();
                }
            })
            .addOnFailureListener(e -> {
                Log.e(TAG, "Text recognition failed: " + e.getMessage(), e);
                hasError.set(true);
                latch.countDown();
            });

        try {
            // Wait for result with timeout
            boolean completed = latch.await(TIMEOUT_MS, TimeUnit.MILLISECONDS);
            if (!completed) {
                Log.w(TAG, "Text recognition timed out after " + TIMEOUT_MS + "ms");
                return new TextRecognitionResult(false, 0.0, 0, 0);
            }
            
            if (hasError.get()) {
                return new TextRecognitionResult(false, 0.0, 0, 0);
            }
            
            TextRecognitionResult result = resultRef.get();
            return result != null ? result : new TextRecognitionResult(false, 0.0, 0, 0);
            
        } catch (InterruptedException e) {
            Log.e(TAG, "Text recognition interrupted: " + e.getMessage(), e);
            Thread.currentThread().interrupt();
            return new TextRecognitionResult(false, 0.0, 0, 0);
        }
    }

    /**
     * Analyze text confidence and determine readability
     * @param visionText Recognized text from ML Kit
     * @return TextRecognitionResult with analysis
     */
    private TextRecognitionResult analyzeTextConfidence(Text visionText) {
        int totalWords = 0;
        int readableWords = 0;
        double totalConfidence = 0.0;
        StringBuilder allText = new StringBuilder();
        StringBuilder readableText = new StringBuilder();
        
        Log.d(TAG, "=== Text Recognition Analysis ===");
        
        for (Text.TextBlock block : visionText.getTextBlocks()) {
            Log.d(TAG, "Text Block: " + block.getText());
            allText.append(block.getText()).append(" ");
            
            for (Text.Line line : block.getLines()) {
                Log.d(TAG, "  Line: " + line.getText());
                
                for (Text.Element element : line.getElements()) {
                    String text = element.getText().trim();
                    if (!text.isEmpty()) {
                        totalWords++;
                        
                        // Get confidence score (ML Kit doesn't provide per-word confidence)
                        // We'll use a heuristic based on text quality
                        double confidence = estimateWordConfidence(element, text);
                        totalConfidence += confidence;
                        
                        Log.d(TAG, String.format("    Word: '%s' | Confidence: %.3f | Readable: %s", 
                                text, confidence, confidence >= MIN_WORD_CONFIDENCE ? "YES" : "NO"));
                        
                        if (confidence >= MIN_WORD_CONFIDENCE) {
                            // Optional dictionary check for enhanced validation
                            if (!useDictionaryCheck || isInDictionary(text)) {
                                readableWords++;
                                readableText.append(text).append(" ");
                                Log.d(TAG, String.format("      ✓ Added to readable words (dict check: %s)", 
                                        useDictionaryCheck ? (isInDictionary(text) ? "PASS" : "FAIL") : "SKIPPED"));
                            } else {
                                Log.d(TAG, "      ✗ Failed dictionary check");
                            }
                        }
                    }
                }
            }
        }
        
        double averageConfidence = totalWords > 0 ? totalConfidence / totalWords : 0.0;
        
            // Image is readable if:
            // 1. We have text AND
            // 2. At least 60% of words are readable OR average confidence is high
            boolean isReadable = totalWords > 0 && 
                               (readableWords >= Math.max(1, totalWords * AT_LEAST_N_PERCENT_OF_WORDS_ARE_READABLE) || averageConfidence >= AT_LEAST_N_PERCENT_OF_AVERAGE_CONFIDENCE);
        
        // Log summary
        Log.d(TAG, "=== Text Recognition Summary ===");
        Log.d(TAG, "All detected text: '" + allText.toString().trim() + "'");
        Log.d(TAG, "Readable text: '" + readableText.toString().trim() + "'");
        Log.d(TAG, String.format("Total words: %d | Readable words: %d | Average confidence: %.3f", 
                totalWords, readableWords, averageConfidence));
        Log.d(TAG, String.format("Readability threshold: %.1f%% (need %.1f%%) | Result: %s", 
                totalWords > 0 ? (readableWords * 100.0 / totalWords) : 0.0,
                AT_LEAST_N_PERCENT_OF_WORDS_ARE_READABLE * 100.0,
                isReadable ? "READABLE" : "BLURRY"));
        Log.d(TAG, "================================");
        
        return new TextRecognitionResult(isReadable, averageConfidence, totalWords, readableWords);
    }

    /**
     * Estimate word confidence based on element properties
     * @param element Text element from ML Kit
     * @param text The actual text content
     * @return Estimated confidence score
     */
    private double estimateWordConfidence(Text.Element element, String text) {
        // Base confidence estimation based on text characteristics
        double confidence = 0.55; // Start with 55% base confidence
        
        // Check text length (very short or very long words might be less reliable)
        if (text.length() >= 3 && text.length() <= 15) {
            confidence += 0.2;
        }
        
        // Check for common patterns that indicate good recognition
        if (text.matches("[a-zA-Z0-9\\s\\-\\.]+")) {
            confidence += 0.15;
        }
        
        // Check for mixed case (indicates good character recognition)
        if (text.matches(".*[a-z].*") && text.matches(".*[A-Z].*")) {
            confidence += 0.1;
        }
        
        // Check for numbers (often well recognized)
        if (text.matches(".*\\d.*")) {
            confidence += 0.1;
        }
        
        // Penalize for special characters that might indicate poor recognition
        if (text.matches(".*[^a-zA-Z0-9\\s\\-\\.].*")) {
            confidence -= 0.1;
        }
        
        return Math.max(0.0, Math.min(1.0, confidence));
    }

    /**
     * Check if word is in common dictionary
     * @param word Word to check
     * @return true if word is in dictionary
     */
    private boolean isInDictionary(String word) {
        if (commonWords == null || word == null) {
            return false;
        }
        
        String cleanWord = word.toLowerCase().replaceAll("[^a-zA-Z]", "");
        return commonWords.contains(cleanWord);
    }

    /**
     * Initialize common English words for dictionary check
     */
    private void initializeCommonWords() {
        commonWords = new HashSet<>();
        
        // Add common English words (sample set - can be expanded)
        String[] words = {
            "the", "be", "to", "of", "and", "a", "in", "that", "have", "i", "it", "for", "not", "on", "with",
            "he", "as", "you", "do", "at", "this", "but", "his", "by", "from", "they", "we", "say", "her",
            "she", "or", "an", "will", "my", "one", "all", "would", "there", "their", "what", "so", "up",
            "out", "if", "about", "who", "get", "which", "go", "me", "when", "make", "can", "like", "time",
            "no", "just", "him", "know", "take", "people", "into", "year", "your", "good", "some", "could",
            "them", "see", "other", "than", "then", "now", "look", "only", "come", "its", "over", "think",
            "also", "back", "after", "use", "two", "how", "our", "work", "first", "well", "way", "even",
            "new", "want", "because", "any", "these", "give", "day", "most", "us", "is", "was", "are",
            "were", "been", "has", "had", "having", "does", "did", "doing", "can", "could", "should",
            "would", "may", "might", "must", "shall", "will", "am", "being", "became", "become", "becomes"
        };
        
        for (String word : words) {
            commonWords.add(word);
        }
    }

    /**
     * Enable or disable dictionary check for enhanced readability detection
     * @param enable true to enable dictionary check
     */
    public void setDictionaryCheckEnabled(boolean enable) {
        this.useDictionaryCheck = enable;
        Log.d(TAG, "Dictionary check " + (enable ? "enabled" : "disabled"));
    }

    /**
     * Check if text recognizer is properly initialized
     * @return true if initialized
     */
    public boolean isInitialized() {
        return isInitialized;
    }

    /**
     * Clean up resources
     */
    public void close() {
        if (textRecognizer != null) {
            textRecognizer.close();
            textRecognizer = null;
        }
        isInitialized = false;
    }

    /**
     * Result class for text recognition analysis
     */
    private static class TextRecognitionResult {
        final boolean isReadable;
        final double averageConfidence;
        final int totalWords;
        final int readableWords;

        TextRecognitionResult(boolean isReadable, double averageConfidence, int totalWords, int readableWords) {
            this.isReadable = isReadable;
            this.averageConfidence = averageConfidence;
            this.totalWords = totalWords;
            this.readableWords = readableWords;
        }
    }
}

File: ios/Plugin/TextRecognitionBlurHelper.swift
import Foundation
import UIKit
import Vision

/**
 * Text Recognition Blur Detection Helper for iOS
 * Uses Vision framework for text recognition to detect blur/sharp images
 * Based on text readability and confidence scores
 */
class TextRecognitionBlurHelper {
    
    private static let TAG = "TextRecognitionBlurHelper"
    private static let MIN_WORD_CONFIDENCE = 0.7 // 70% confidence threshold
    private static let AT_LEAST_N_PERCENT_OF_WORDS_ARE_READABLE = 0.6 // 60% of words are readable
    private static let AT_LEAST_N_PERCENT_OF_AVERAGE_CONFIDENCE = 0.85 // 85% of average confidence
    
    private var isInitialized = false
    private var commonWords: Set<String>
    private var useDictionaryCheck = false
    
    // Text recognition result structure
    private struct TextRecognitionResult {
        let isReadable: Bool
        let averageConfidence: Double
        let totalWords: Int
        let readableWords: Int
    }
    
    init() {
        // Initialize common English words for dictionary check
        commonWords = Set<String>()
        initializeCommonWords()
    }
    
    /**
     * Initialize the text recognition helper
     * @return true if initialization successful
     */
    func initialize() -> Bool {
        // Vision framework is available by default on iOS, no specific initialization needed
        isInitialized = true
        return true
    }
    
    /**
     * Detect if image is blurry based on text recognition confidence
     * @param image Input UIImage
     * @return true if image is blurry (text not readable), false if sharp
     */
    func isBlurry(image: UIImage) -> Bool {
        if !isInitialized {
            return false // Default to not blurry if not initialized
        }
        
        
        do {
            let result = try detectTextWithConfidence(image: image)
            
            let isBlurry = !result.isReadable
            
            return isBlurry
        } catch {
            return false // Default to not blurry on error
        }
    }
    
    /**
     * Detect blur with detailed confidence scores
     * @param image Input UIImage
     * @return Dictionary with isBlur, textConfidence, wordCount, and readableWords
     */
    func detectBlurWithConfidence(image: UIImage) -> [String: Any] {
        var result: [String: Any] = [:]
        
        if !isInitialized {
            result["isBlur"] = false
            result["textConfidence"] = 0.0
            result["wordCount"] = 0
            result["readableWords"] = 0
            result["hasText"] = false
            result["error"] = "Text recognition helper not initialized"
            return result
        }
        
        do {
            let recognitionResult = try detectTextWithConfidence(image: image)
            
            result["isBlur"] = !recognitionResult.isReadable
            result["textConfidence"] = recognitionResult.averageConfidence
            result["wordCount"] = recognitionResult.totalWords
            result["readableWords"] = recognitionResult.readableWords
            result["hasText"] = recognitionResult.totalWords > 0
            
            
            return result
            
        } catch {
            result["isBlur"] = false
            result["textConfidence"] = 0.0
            result["wordCount"] = 0
            result["readableWords"] = 0
            result["hasText"] = false
            result["error"] = error.localizedDescription
            return result
        }
    }
    
    /**
     * Detect text in image with confidence analysis using Vision framework
     * @param image Input UIImage
     * @return TextRecognitionResult with confidence and readability info
     */
    private func detectTextWithConfidence(image: UIImage) throws -> TextRecognitionResult {
        guard let cgImage = image.cgImage else {
            throw NSError(domain: "TextRecognitionError", code: 1, userInfo: [NSLocalizedDescriptionKey: "Failed to get CGImage from UIImage"])
        }
        
        let requestHandler = VNImageRequestHandler(cgImage: cgImage, options: [:])
        
        var recognitionResult: TextRecognitionResult?
        var recognitionError: Error?
        
        // Use semaphore for synchronous execution
        let semaphore = DispatchSemaphore(value: 0)
        
        let request = VNRecognizeTextRequest { [weak self] (request, error) in
            defer { semaphore.signal() }
            
            if let error = error {
                recognitionError = error
                return
            }
            
            guard let observations = request.results as? [VNRecognizedTextObservation] else {
                recognitionResult = TextRecognitionResult(isReadable: false, averageConfidence: 0.0, totalWords: 0, readableWords: 0)
                return
            }
            
            recognitionResult = self?.analyzeTextConfidence(observations: observations)
        }
        
        // Configure text recognition for better accuracy
        request.recognitionLevel = .accurate
        request.usesLanguageCorrection = true
        
        // Perform text recognition
        try requestHandler.perform([request])
        
        // Wait for completion with timeout
        let timeout = DispatchTime.now() + .seconds(5)
        if semaphore.wait(timeout: timeout) == .timedOut {
            throw NSError(domain: "TextRecognitionError", code: 2, userInfo: [NSLocalizedDescriptionKey: "Text recognition timed out"])
        }
        
        if let error = recognitionError {
            throw error
        }
        
        return recognitionResult ?? TextRecognitionResult(isReadable: false, averageConfidence: 0.0, totalWords: 0, readableWords: 0)
    }
    
    /**
     * Analyze text confidence and determine readability
     * @param observations Recognized text observations from Vision framework
     * @return TextRecognitionResult with analysis
     */
    private func analyzeTextConfidence(observations: [VNRecognizedTextObservation]) -> TextRecognitionResult {
        var totalWords = 0
        var readableWords = 0
        var totalConfidence = 0.0
        var allText = ""
        var readableText = ""
        
        
        for observation in observations {
            // Get the top candidate for each observation
            guard let topCandidate = observation.topCandidates(1).first else { continue }
            
            let text = topCandidate.string
            let visionConfidence = Double(topCandidate.confidence)
            
            allText += text + " "
            
            // Split text into words for individual analysis
            let words = text.components(separatedBy: .whitespacesAndNewlines).filter { !$0.isEmpty }
            
            for word in words {
                let trimmedWord = word.trimmingCharacters(in: .punctuationCharacters)
                if !trimmedWord.isEmpty {
                    totalWords += 1
                    
                    // Estimate word confidence based on Vision confidence and text characteristics
                    let wordConfidence = estimateWordConfidence(word: trimmedWord, visionConfidence: visionConfidence)
                    totalConfidence += wordConfidence
                    
                    
                    if wordConfidence >= Self.MIN_WORD_CONFIDENCE {
                        // Optional dictionary check for enhanced validation
                        if !useDictionaryCheck || isInDictionary(word: trimmedWord) {
                            readableWords += 1
                            readableText += trimmedWord + " "
                        } else {
                        }
                    }
                }
            }
        }
        
        let averageConfidence = totalWords > 0 ? totalConfidence / Double(totalWords) : 0.0
        
        // Image is readable if:
        // 1. We have text AND
        // 2. At least 60% of words are readable OR average confidence is high
        let isReadable = totalWords > 0 &&
                        (readableWords >= max(1, Int(Double(totalWords) * Self.AT_LEAST_N_PERCENT_OF_WORDS_ARE_READABLE)) ||
                         averageConfidence >= Self.AT_LEAST_N_PERCENT_OF_AVERAGE_CONFIDENCE)
        
        
        return TextRecognitionResult(isReadable: isReadable, averageConfidence: averageConfidence, totalWords: totalWords, readableWords: readableWords)
    }
    
    /**
     * Estimate word confidence based on Vision confidence and text characteristics
     * @param word The actual word text
     * @param visionConfidence Confidence from Vision framework
     * @return Estimated confidence score
     */
    private func estimateWordConfidence(word: String, visionConfidence: Double) -> Double {
        // Start with Vision framework confidence
        var confidence = visionConfidence
        
        // Apply text characteristic adjustments (similar to Android implementation)
        
        // Check text length (very short or very long words might be less reliable)
        if word.count >= 3 && word.count <= 15 {
            confidence += 0.1 // Boost confidence for reasonable length words
        }
        
        // Check for common patterns that indicate good recognition
        let alphanumericPattern = "^[a-zA-Z0-9\\s\\-\\.]+$"
        if word.range(of: alphanumericPattern, options: .regularExpression) != nil {
            confidence += 0.05
        }
        
        // Check for mixed case (indicates good character recognition)
        let hasLowercase = word.range(of: "[a-z]", options: .regularExpression) != nil
        let hasUppercase = word.range(of: "[A-Z]", options: .regularExpression) != nil
        if hasLowercase && hasUppercase {
            confidence += 0.05
        }
        
        // Check for numbers (often well recognized)
        if word.range(of: "\\d", options: .regularExpression) != nil {
            confidence += 0.05
        }
        
        // Penalize for special characters that might indicate poor recognition
        let specialCharPattern = "[^a-zA-Z0-9\\s\\-\\.]"
        if word.range(of: specialCharPattern, options: .regularExpression) != nil {
            confidence -= 0.05
        }
        
        return max(0.0, min(1.0, confidence))
    }
    
    /**
     * Check if word is in common dictionary
     * @param word Word to check
     * @return true if word is in dictionary
     */
    private func isInDictionary(word: String) -> Bool {
        guard !commonWords.isEmpty else { return false }
        
        let cleanWord = word.lowercased().replacingOccurrences(of: "[^a-zA-Z]", with: "", options: .regularExpression)
        return commonWords.contains(cleanWord)
    }
    
    /**
     * Initialize common English words for dictionary check
     */
    private func initializeCommonWords() {
        let words = [
            "the", "be", "to", "of", "and", "a", "in", "that", "have", "i", "it", "for", "not", "on", "with",
            "he", "as", "you", "do", "at", "this", "but", "his", "by", "from", "they", "we", "say", "her",
            "she", "or", "an", "will", "my", "one", "all", "would", "there", "their", "what", "so", "up",
            "out", "if", "about", "who", "get", "which", "go", "me", "when", "make", "can", "like", "time",
            "no", "just", "him", "know", "take", "people", "into", "year", "your", "good", "some", "could",
            "them", "see", "other", "than", "then", "now", "look", "only", "come", "its", "over", "think",
            "also", "back", "after", "use", "two", "how", "our", "work", "first", "well", "way", "even",
            "new", "want", "because", "any", "these", "give", "day", "most", "us", "is", "was", "are",
            "were", "been", "has", "had", "having", "does", "did", "doing", "can", "could", "should",
            "would", "may", "might", "must", "shall", "will", "am", "being", "became", "become", "becomes"
        ]
        
        commonWords = Set(words)
    }
    
    /**
     * Enable or disable dictionary check for enhanced readability detection
     * @param enable true to enable dictionary check
     */
    func setDictionaryCheckEnabled(_ enable: Bool) {
        useDictionaryCheck = enable
    }
    
    /**
     * Check if text recognition helper is properly initialized
     * @return true if initialized
     */
    func getIsInitialized() -> Bool {
        return isInitialized
    }
    
    /**
     * Clean up resources
     */
    func close() {
        isInitialized = false
    }
}
